src/NHibernate/Linq/ReWriters/RemoveUnnecessaryBodyOperators.cs

```diff
@@ -21,20 +22,32 @@ public static void ReWrite(QueryModel queryModel)
 
 		internal static void RemoveUnnecessaryOrderByClauses(QueryModel queryModel)
 		{
-			if (queryModel.ResultOperators.Count == 1 &&
-			    queryModel.ResultOperators.All(
-				    r => r is ContainsResultOperator || r is AnyResultOperator || r is AllResultOperator))
+			if (IsOrderByNeeded(queryModel))
+				return;
+
+			// For these operators, we can remove any order-by clause
+			var bodyClauses = queryModel.BodyClauses;
+			for (int i = bodyClauses.Count - 1; i >= 0; i--)
 			{
-				// For these operators, we can remove any order-by clause
-				var bodyClauses = queryModel.BodyClauses;
-				for (int i = bodyClauses.Count - 1; i >= 0; i--)
-				{
-					if (bodyClauses[i] is OrderByClause)
-						bodyClauses.RemoveAt(i);
-				}
+				if (bodyClauses[i] is OrderByClause)
+					bodyClauses.RemoveAt(i);
 			}
 		}
 
+		internal static bool IsOrderByNeeded(QueryModel queryModel)
+		{
+			switch (queryModel.ResultOperators.Count)
+			{
+				case 1:
+					var r = queryModel.ResultOperators[0];
+					return !(r is AnyResultOperator || r is AllResultOperator || r is ContainsResultOperator);  
+				case 0:
+					return !(queryModel.SelectClause.Selector is NhCountExpression);
```

Comment:
```suggestion
					return !(queryModel.SelectClause.Selector is NhAggregatedExpression);
```

Reply:
But... distinct would probably require order-by.

Reply:
It's not required but we shouldn't loose it